File: smoke_tests/swift/sensitive_data.swift
// Sensitive Data Exposure vulnerabilities in Swift
import Foundation

class SensitiveDataExposure {

    // Test 1: Hardcoded API key
    func makeApiCall() {
        // VULNERABLE: Hardcoded API key
        let apiKey = "sk-live-abcd1234567890xyz"
        var request = URLRequest(url: URL(string: "https://api.example.com")!)
        request.setValue(apiKey, forHTTPHeaderField: "Authorization")
    }

    // Test 2: Credentials in source code
    func connectToDatabase() {
        // VULNERABLE: Hardcoded credentials
        let username = "admin"
        let password = "SuperSecret123!"
        let connectionString = "mysql://\(username):\(password)@db.example.com/production"
    }

    // Test 3: Private key in code
    func signData(data: Data) {
        // VULNERABLE: Private key embedded
        let privateKey = """
        -----BEGIN RSA PRIVATE KEY-----
        MIIEpAIBAAKCAQEA0Z3VS5JJcds3xfn/ygWyF8PbnGy...
        -----END RSA PRIVATE KEY-----
        """
    }

    // Test 4: Sensitive data in URL
    func fetchUserData(userId: String, ssn: String) {
        // VULNERABLE: SSN in URL
        let url = URL(string: "https://api.example.com/user/\(userId)?ssn=\(ssn)")!
        URLSession.shared.dataTask(with: url).resume()
    }

    // Test 5: Credit card in logs
    func processPayment(cardNumber: String, cvv: String) {
        // VULNERABLE: Card data logged
        print("Processing card: \(cardNumber)")
        NSLog("CVV: %@", cvv)
    }

    // Test 6: PII stored unencrypted
    func storePii(ssn: String, dob: String) {
        // VULNERABLE: Sensitive data in UserDefaults
        UserDefaults.standard.set(ssn, forKey: "user_ssn")
        UserDefaults.standard.set(dob, forKey: "user_dob")
    }

    // Test 7: Password in error message
    func validateCredentials(username: String, password: String) throws {
        // VULNERABLE: Password in error
        if !isValidPassword(password) {
            throw NSError(domain: "Auth", code: 1,
                         userInfo: [NSLocalizedDescriptionKey: "Invalid password: \(password)"])
        }
    }

    // Test 8: Token exposure in response
    func handleAuthResponse() -> [String: Any] {
        let token = generateToken()
        let refreshToken = generateRefreshToken()
        // VULNERABLE: Exposing all tokens
        return [
            "access_token": token,
            "refresh_token": refreshToken,
            "token_type": "Bearer"
        ]
    }

    // Test 9: Health data without encryption
    func storeHealthRecord(data: HealthRecord) {
        // VULNERABLE: Health data stored plaintext
        let encoder = JSONEncoder()
        if let encoded = try? encoder.encode(data) {
            try? encoded.write(to: URL(fileURLWithPath: "/data/health.json"))
        }
    }

    // Test 10: Biometric template stored insecurely
    func storeBiometricTemplate(template: Data) {
        // VULNERABLE: Biometric data in plaintext
        UserDefaults.standard.set(template, forKey: "biometric_template")
    }

    // Test 11: Session token in cookie without secure flag
    func setSessionCookie(token: String) -> String {
        // VULNERABLE: Missing Secure and HttpOnly flags
        return "session=\(token); Path=/"
    }

    // Test 12: AWS credentials
    func configureAws() {
        // VULNERABLE: AWS credentials in code
        let accessKey = "AKIAIOSFODNN7EXAMPLE"
        let secretKey = "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
    }

    // Test 13: Encryption key hardcoded
    func encryptData(data: Data) -> Data? {
        // VULNERABLE: Hardcoded encryption key
        let key = "MySecretEncryptionKey123"
        return encrypt(data: data, key: key)
    }

    // Test 14: Backup with sensitive data
    func backupUserData() {
        let userData = [
            "email": "user@example.com",
            "password": "userPassword123",  // VULNERABLE
            "creditCard": "4111111111111111"  // VULNERABLE
        ]
        // Backup to iCloud
    }

    private func isValidPassword(_ password: String) -> Bool { true }
    private func generateToken() -> String { "token" }
    private func generateRefreshToken() -> String { "refresh" }
    private func encrypt(data: Data, key: String) -> Data? { nil }
}

struct HealthRecord: Codable {
    var patientId: String
    var diagnosis: String
    var medications: [String]
}

File: smoke_tests/dart/sensitive_data.dart
// Sensitive Data Exposure vulnerabilities in Dart

import 'dart:io';
import 'dart:convert';

// Test 1: Logging sensitive data
void vulnerableLogPassword(String username, String password) {
  // VULNERABLE: Password in logs
  print('Login attempt: $username / $password');
}

// Test 2: Credit card in logs
void vulnerableLogCreditCard(String cardNumber, String cvv) {
  // VULNERABLE: Credit card data in logs
  print('Processing payment: $cardNumber, CVV: $cvv');
}

// Test 3: API key in error message
void vulnerableApiKeyInError(String apiKey, Exception error) {
  // VULNERABLE: API key exposed in error
  throw Exception('API call failed with key $apiKey: $error');
}

// Test 4: Sensitive data in URL
String vulnerableSensitiveInUrl(String ssn) {
  // VULNERABLE: SSN in URL
  return 'https://api.example.com/verify?ssn=$ssn';
}

// Test 5: PII in response
Map<String, dynamic> vulnerablePiiInResponse(Map<String, dynamic> user) {
  // VULNERABLE: Returning sensitive fields
  return {
    'id': user['id'],
    'name': user['name'],
    'ssn': user['ssn'],
    'password_hash': user['password_hash'],
    'credit_card': user['credit_card'],
  };
}

// Test 6: Sensitive data in exception
void vulnerableSensitiveException(String token) {
  // VULNERABLE: Token in exception message
  throw StateError('Invalid authentication token: $token');
}

// Test 7: Debug mode exposing data
void vulnerableDebugMode(HttpRequest request, HttpResponse response, Object error) {
  // VULNERABLE: Full error details in production
  var debug = true;
  if (debug) {
    response.write(jsonEncode({
      'error': error.toString(),
      'stack': StackTrace.current.toString(),
      'request': {
        'headers': request.headers.toString(),
        'cookies': request.cookies.map((c) => '${c.name}=${c.value}').toList(),
      }
    }));
  }
}

// Test 8: Verbose error messages
void vulnerableVerboseError(HttpResponse response, String query, Exception e) {
  // VULNERABLE: SQL query in error response
  response.write('Database error executing: $query - ${e.toString()}');
}

// Test 9: Sensitive data in comments
class VulnerableComments {
  // Default admin password: admin123
  // API Key: sk_live_abc123xyz
  void authenticate() {}
}

// Test 10: Hardcoded test credentials
class VulnerableTestCreds {
  // VULNERABLE: Test credentials in code
  static const testUser = 'admin@example.com';
  static const testPassword = 'Password123!';
  static const testApiKey = 'test_api_key_12345';
}

// Test 11: Sensitive data in temp file
Future<void> vulnerableTempFile(String creditCard) async {
  // VULNERABLE: Sensitive data in temp file
  var tempFile = File('/tmp/payment_${DateTime.now().millisecondsSinceEpoch}.txt');
  await tempFile.writeAsString('Card: $creditCard');
}

// Test 12: Caching sensitive data
class VulnerableCache {
  static final Map<String, String> cache = {};

  // VULNERABLE: Caching sensitive data in memory
  static void cacheUserData(String id, String ssn, String password) {
    cache['user_$id'] = jsonEncode({'ssn': ssn, 'password': password});
  }
}

// Test 13: Sensitive data in analytics
void vulnerableAnalytics(String userId, String email, String ipAddress) {
  // VULNERABLE: PII in analytics
  analytics.track('user_login', {
    'user_id': userId,
    'email': email,
    'ip': ipAddress,
  });
}

// Test 14: Autocomplete on sensitive fields
String vulnerableAutocomplete() {
  // VULNERABLE: Autocomplete enabled on sensitive fields
  return '''
    <input type="text" name="ssn" autocomplete="on">
    <input type="password" name="password" autocomplete="on">
  ''';
}

// Test 15: Sensitive data in local storage
void vulnerableLocalStorage(String token) {
  // VULNERABLE: Token in local storage (web context)
  window.localStorage['auth_token'] = token;
}

// Test 16: Sensitive headers logged
void vulnerableLogHeaders(HttpRequest request) {
  // VULNERABLE: Logging authorization header
  print('Headers: ${request.headers}');
  print('Auth: ${request.headers.value("Authorization")}');
}

// Test 17: Returning stack trace to client
void vulnerableStackTrace(HttpResponse response, Exception e, StackTrace stack) {
  // VULNERABLE: Stack trace to client
  response.write(jsonEncode({
    'error': e.toString(),
    'trace': stack.toString(),
  }));
}

// Test 18: Sensitive data in form action
String vulnerableFormAction(String accountId) {
  // VULNERABLE: Account ID in form action URL
  return '<form action="/api/account/$accountId/update" method="POST">';
}

// Test 19: Session data exposure
void vulnerableSessionExposure(HttpResponse response, Map<String, dynamic> session) {
  // VULNERABLE: Full session data in response
  response.write(jsonEncode({'session': session}));
}

// Test 20: Backup with sensitive data
Future<void> vulnerableBackup(Map<String, dynamic> userData) async {
  // VULNERABLE: Unencrypted backup of sensitive data
  var backupFile = File('/backups/user_data.json');
  await backupFile.writeAsString(jsonEncode(userData));
}

// Mock classes
class Analytics {
  void track(String event, Map<String, dynamic> properties) {}
}

class Window {
  final Map<String, String> localStorage = {};
}

final analytics = Analytics();
final window = Window();
